src/changestats.bnp4c.c

```diff
@@ -0,0 +1,312 @@
+     EstimNetDirected or statnet gwesp etc. cache */
+
+  /* In an undirected network, each edge is only stored as (tail, head) where
+     tail < head, so to step through all edges of a node it is necessary
+     to step through all outedges and also through all inedges */
+  STEP_THROUGH_OUTEDGES(i, edge1, vnode) {     /* i -- v */
+    if (vnode == i || vnode == j)
+      continue;
+    STEP_THROUGH_OUTEDGES(j, edge2, wnode) {
+      if (wnode == vnode)     /* v -- j */
+        count++;
+    }
+    STEP_THROUGH_INEDGES(j, edge2, wnode) {
+      if (wnode == vnode)      /* v -- j */
+        count++;
+    }
+  }
+  STEP_THROUGH_INEDGES(i, edge1, vnode) {     /* i -- v */
+    if (vnode == i || vnode == j)
+      continue;
+    STEP_THROUGH_OUTEDGES(j, edge2, wnode) {
+      if (wnode == vnode)     /* v -- j */
+        count++;
+    }
+    STEP_THROUGH_INEDGES(j, edge2, wnode) {
+      if (wnode == vnode)      /* v -- j */
+        count++;
+    }
+  }
+  return count;
+}
+
+
+
+/*
+ * number of four-cycles that node unode is involved in
+ *
+ */
+static unsigned long num_fourcycles_node(Network *nwp, Vertex unode)  {
+  /* Note Network *nwp parameter has to be called nwp for use of macros */
+
+  Vertex vnode, wnode;
+  Edge edge1, edge2;
+  int *visited = R_Calloc(N_NODES, int); /* list of visited nodes */
+  unsigned long fourcycle_count = 0;
+
+  /*
+    Note it seems that the Vertex is an int from 1 .. N_NODES
+     (R indexing from 1 rather than C indexing from 0)
+     so we subtract one when indexing the visited array.
+     This code depends on this being the case.
+  */
+  
+  /* this involves iterating over all nodes that are distance 2 from unode */
+  /* In an undirected network, each edge is only stored as (tail, head) where
+     tail < head, so to step through all edges of a node it is necessary
+     to step through all outedges and also through all inedges */
+  STEP_THROUGH_OUTEDGES(unode, edge1, vnode) {
+    STEP_THROUGH_OUTEDGES(vnode, edge2, wnode) {
+      if (wnode != unode && !visited[wnode-1]) {
+        visited[wnode-1] = 1;
+        fourcycle_count += n_choose_2(twopaths(nwp, unode, wnode));
+      }
+    }
+    STEP_THROUGH_INEDGES(vnode, edge2, wnode) {
+      if (wnode != unode && !visited[wnode-1]) {
+        visited[wnode-1] = 1;
+        fourcycle_count += n_choose_2(twopaths(nwp, unode, wnode));
+      }
+    }
+  }
+  STEP_THROUGH_INEDGES(unode, edge1, vnode) {
+    STEP_THROUGH_OUTEDGES(vnode, edge2, wnode) {
+      if (wnode != unode && !visited[wnode-1]) {
+        visited[wnode-1] = 1;
+        fourcycle_count += n_choose_2(twopaths(nwp, unode, wnode));
+      }
+    }
+    STEP_THROUGH_INEDGES(vnode, edge2, wnode) {
+      if (wnode != unode && !visited[wnode-1]) {
+        visited[wnode-1] = 1;
+        fourcycle_count += n_choose_2(twopaths(nwp, unode, wnode));
+      }
+    }
+  }
+  R_Free(visited);
+  return fourcycle_count;
+}
```

Comment:
1. Why not cache this with the storage API?
2. Allocating and freeing memory each time a change statistic is calculated is costly; it would be better to allocate in an initialiser and free in the finaliser.